File: 0152-maximum-product-subarray/solution.cpp
class Solution {
public:
    int maxProduct(vector<int>& n) {
        int r = n[0], N = n.size();
        for(int i = 1, imax = r, imin = r; i < N; i++) {
            if (n[i] < 0) swap(imax, imin);
            imax = max(n[i], imax * n[i]);
            imin = min(n[i], imin * n[i]);
            r = max(r, imax);
        }
        return r;
    }
};

File: 0152-maximum-product-subarray/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]

        currMin, currMax = 1, 1

        for n in nums:
            tmp = currMax * n
            currMax = max(n * currMax, n * currMin, n)
            currMin = min(tmp, n *currMin, n)
            res = max(res, currMax)

        return res

